src/helpers/parseUtil.ts

```diff
@@ -109,17 +111,76 @@ export type ParseParams = {
   async: boolean;
 };
 
-export type ParseParamsWithOptionals = util.flatten<
-  Partial<ParseParams> & { data: any }
->;
-
 export type ParseParamsNoData = Omit<ParseParams, "data">;
 
-export type ParseContext = ParseParams &
-  ReturnType<typeof issueHelpers> & {
-    parsedType: ZodParsedType;
-    currentError: ZodError;
-  };
+export type ParsePathComponent = string | number;
+
+export type ParsePath = null | {
+  readonly component: ParsePathComponent;
+  readonly parent: ParsePath;
+  readonly count: number;
+};
+
+export const EMPTY_PATH: ParsePath = null;
+
+export const pathToArray = (path: ParsePath): ParsePathComponent[] => {
+  if (path === null) return [];
+  const arr: ParsePathComponent[] = new Array(path.count);
+  while (path !== null) {
+    arr[path.count - 1] = path.component;
+    path = path.parent;
+  }
+  return arr;
+};
+
+export const pathFromArray = (arr: ParsePathComponent[]): ParsePath => {
+  let path: ParsePath = null;
+  for (let i = 0; i < arr.length; i++) {
+    path = { parent: path, component: arr[i], count: i + 1 };
+  }
+  return path;
+};
+
+export type ParseContextParameters = {
+  errorMap: ZodErrorMap;
+  async: boolean;
+};
+
+export class ParseContext {
+  constructor(
+    public readonly path: ParsePath,
+    public readonly issues: ZodIssue[],
+    public readonly params: ParseContextParameters
+  ) {}
```

Comment:
Yes, the reason was purely because of performance: adding `parsedType` and `data` required a shallow clone for every level and also accessing them through the object property which had (a slight) penalty hit compared to passing them as separate parameters. Of course there is still the `path` that changes quite frequently so if we want to maximize the performance, that should be passed as a parameter as well.

Reply:
Seems fine to leave that as a future optimization since it wouldn't affect the public API of the library.